notification/notifier.go

```diff
@@ -0,0 +1,433 @@
+	if err != nil {
+		log.Error("Failed to deserialize blockHeader in new block notification: "+
+			"%v", err)
+		return
+	}
+	height := int32(blockHeader.Height)
+	hash := blockHeader.BlockHash()
+	prevHash := blockHeader.PrevBlock // to ensure this is the next block
+
+	log.Tracef("OnBlockConnected: %d / %v (previous: %v)", height, hash, prevHash)
+
+	notifier.anyQ <- blockHeader
+}
+
+// rpcclient.NotificationHandlers.OnBlockDisconnected
+func (notifier *Notifier) onBlockDisconnected(blockHeaderSerialized []byte) {
+	blockHeader := new(wire.BlockHeader)
+	err := blockHeader.FromBytes(blockHeaderSerialized)
+	if err != nil {
+		log.Error("Failed to deserialize blockHeader in block disconnect notification: "+
+			"%v", err)
+		return
+	}
+	height := int32(blockHeader.Height)
+	hash := blockHeader.BlockHash()
+
+	log.Tracef("OnBlockDisconnected: %d / %v", height, hash)
+}
+
+// rpcclient.NotificationHandlers.OnReorganization
+// Processes the arguments into a BranchTips before submitting to the anyQ.
+func (notifier *Notifier) onReorganization(oldHash *chainhash.Hash, oldHeight int32,
+	newHash *chainhash.Hash, newHeight int32) {
+	log.Tracef("OnReorganization: %d / %s --> %d / %s",
+		oldHeight, oldHash, newHeight, newHash)
+
+	notifier.anyQ <- BranchTips{
+		OldChainHead:   *oldHash,
+		OldChainHeight: oldHeight,
+		NewChainHead:   *newHash,
+		NewChainHeight: newHeight,
+	}
+}
+
+// rpcclient.NotificationHandlers.OnWinningTickets
+func (notifier *Notifier) onWinningTickets(_ *chainhash.Hash, _ int64, tickets []*chainhash.Hash) {
+	var txstr []string
+	for _, t := range tickets {
+		txstr = append(txstr, t.String())
+	}
+	log.Tracef("Winning tickets: %v", strings.Join(txstr, ", "))
+}
+
+// rpcclient.NotificationHandlers.OnNewTickets
+func (notifier *Notifier) onNewTickets(_ *chainhash.Hash, _ int64, _ int64, tickets []*chainhash.Hash) {
+	for _, tick := range tickets {
+		log.Tracef("Mined new ticket: %v", tick.String())
+	}
+}
+
+// rpcclient.NotificationHandlers.OnTxAcceptedVerbose
+func (notifier *Notifier) onTxAcceptedVerbose(tx *dcrjson.TxRawResult) {
+	// Current UNIX time to assign the new transaction.
+	tx.Time = time.Now().Unix()
+	notifier.anyQ <- tx
+}
+
+// RegisterTxHandlers adds a group of tx handlers. Groups are run sequentially
+// in the order they are registered, but the handlers within the group are run
+// asynchronously.
+func (notifier *Notifier) RegisterTxHandlers(handlers ...TxHandler) {
+	notifier.tx = append(notifier.tx, handlers)
+}
+
+// RegisterBlockHandlers adds a group of block handlers. Groups are run
+// sequentially in the order they are registered, but the handlers within the
+// group are run asynchronously. Handlers registered with RegisterBlockHandlers
+// are FIFO'd together with handlers registered with RegisterBlockHandlersLite.
+func (notifier *Notifier) RegisterBlockHandlers(handlers ...BlockHandler) {
+	notifier.block = append(notifier.block, handlers)
```

Comment:
I have some suggestions for the register functions because I scanned through main.go first and was surprised to see:

```go
notifier.RegisterBlockHandlers(sdbChainMonitor.ConnectBlock)
notifier.RegisterBlockHandlers(wsChainMonitor.ConnectBlock)
```
My thought was, why not just `notifier.RegisterBlockHandlers(sdbChainMonitor.ConnectBlock, wsChainMonitor.ConnectBlock)`, and the answer (because they should not be run concurrently) was not clear until I looked at how `(*Notifier).processBlock` is implemented.  That made it quite clear that there is the concept of a group, the handlers of which are executed concurrently, and that there are multiple groups.

Based on the semantics of the `Register*` functions and how the handlers (and handler groups) are executed in `(*Notifier).processBlock`, I'm thinking about some possible changes:
- The argument.  `RegisterBlockHandlers(handlers []BlockHandler)` Makes it a little more clear that the handlers being registered are a group.  Requires more syntax from the caller.
- The function name. `RegisterConcurrentBlockHandlers` to make it explicit that the handlers in the group provided are executed concurrently.
- A new type for the inputs, `type ConcurrentHandlers []BlockHandler`.
- Mixing the above.  e.g.

```go
func (notifier *Notifier) RegisterBlockHandlerGroups(handlerGroups ...ConcurrentHandlers) {
	notifier.block = append(notifier.block, handlerGroups...)
}
```

making in the complex case of reorg handlers like:

```go
// type ConcurrentReorgHandlers []ReorgHandler

notifier.RegisterReorgHandlerGroups(
	notify.ConcurrentReorgHandlers{sdbChainMonitor.ReorgHandler},
	notify.ConcurrentReorgHandlers{WiredDBChainMonitor.ReorgHandler,
		wsChainMonitor.ReorgHandler, pgDBChainMonitor.ReorgHandler}),
	notify.ConcurrentReorgHandlers{charts.ReorgHandler})
```

But perhaps that's more complex and confusing.  Perhaps just extra comments in main.go where the groups are registered, and at the function definitions would do it.